jujugui/static/gui/src/app/store/env/controller-api.js

```diff
@@ -0,0 +1,683 @@
+        var facadeList = response.facades || [];
+        var facades = facadeList.reduce(function(previous, current) {
+          previous[current.name] = current.versions;
+          return previous;
+        }, {});
+        this.set('facades', facades);
+        var userInfo = response['user-info'];
+        this.set('readOnly', !!userInfo['read-only']);
+        this.set('serverTag', response['server-tag']);
+        // Start pinging the server.
+        // XXX frankban: this is only required as a temporary workaround to
+        // prevent Apache to disconnect the WebSocket in the embedded Juju.
+        if (!this._pinger) {
+          this._pinger = setInterval(
+            this.ping.bind(this), PING_INTERVAL * 1000);
+        }
+        // Clean up for log out text.
+        this.failedAuthentication = false;
+      } else {
+        // If the credentials were rejected remove them.
+        this.setCredentials(null);
+        this.failedAuthentication = true;
+      }
+      this.fire('login', {data: {
+        result: this.userIsAuthenticated,
+        error: data.error || null
+      }});
+    },
+
+    /**
+      Return a version for the given facade name which is supported by the
+      current Juju controller. If a version is provided, return the version
+      number itself if supported, or null if that specific version is not
+      served by the controller. Otherwise, if no version is specified, return
+      the most recent supported version or null if the facade is not found.
+
+      @method findFacadeVersion
+      @param {String} name The facade name (for instance "Application").
+      @param {Int} version The optional facade version (for instance 1 or 2).
+      @return {Int} The facade version or null if facade is not supported.
+    */
+    findFacadeVersion: function(name, version) {
+      var facades = this.get('facades') || {};
+      var versions = facades[name] || [];
+      if (!versions.length) {
+        return null;
+      }
+      if (version === undefined || version === null) {
+        return versions[versions.length - 1];
+      }
+      if (versions.indexOf(version) > -1) {
+        return version;
+      }
+      return null;
+    },
+
+    /**
+     * Attempt to log the user in.  Credentials must have been previously
+     * stored on the environment.
+     *
+     * @method login
+     * @return {undefined} Nothing.
+     */
+    login: function() {
+      // If the user is already authenticated there is nothing to do.
+      if (this.userIsAuthenticated) {
+        this.fire('login', {data: {result: true}});
+        return;
+      }
+      if (this.pendingLoginResponse) {
+        return;
+      }
+      var credentials = this.getCredentials();
+      if (!credentials.user || !credentials.password) {
+        console.warn('attempted login without providing credentials');
+        this.fire('login', {data: {result: false}});
+        return;
+      }
+      this._send_rpc({
```

Comment:
Not here, but in the corresponding method of api.js, please add a TODO for handling the RedirectError response from login, possibly returned by jimm... and create a card.